accord-core/src/main/java/accord/impl/SimpleProgressLog.java

```diff
@@ -668,242 +710,123 @@ State recordExecute(TxnId txnId)
-        for (Instance instance : instances)
-        {
-            // TODO: we want to be able to poll others about pending dependencies to check forward progress,
-            //       as we don't know all dependencies locally (or perhaps any, at execution time) so we may
-            //       begin expecting forward progress too early
-            // state map may be updated during iteration, so need to clone the values set
-            new ArrayList<>(instance.stateMap.values()).forEach(state -> state.update(node));
-        }
-    }
-
-    static class CoordinateApplyAndCheck extends AsyncFuture<Void> implements Callback<ApplyAndCheckOk>
-    {
-        final TxnId txnId;
-        final HomeState state;
-        final Set<Id> waitingOnResponses;
-
-        static Future<Void> applyAndCheck(Node node, TxnId txnId, Command command, HomeState state)
-        {
-            CoordinateApplyAndCheck coordinate = new CoordinateApplyAndCheck(txnId, state);
-            Topologies topologies = node.topology().preciseEpochs(command.txn(), command.executeAt().epoch);
-            state.globalNotPersisted.retainAll(topologies.nodes()); // we might have had some nodes from older shards that are now redundant
-            node.send(state.globalNotPersisted, id -> new ApplyAndCheck(id, topologies,
-                                                                        command.txnId(), command.txn(), command.homeKey(),
-                                                                        command.savedDeps(), command.executeAt(),
-                                                                        command.writes(), command.result(),
-                                                                        state.globalNotPersisted),
-                      coordinate);
-            return coordinate;
-        }
-
-        CoordinateApplyAndCheck(TxnId txnId, HomeState state)
+        public void durableLocal(TxnId txnId)
         {
-            this.txnId = txnId;
-            this.state = state;
-            this.waitingOnResponses = new HashSet<>(state.globalNotPersisted);
+            State state = ensure(txnId);
+            state.global().durableLocal(node);
         }
 
         @Override
-        public void onSuccess(Id from, ApplyAndCheckOk response)
+        public void durable(Command command, @Nullable Set<Id> persistedOn)
         {
-            state.globalNotPersisted.retainAll(response.notPersisted);
-            state.refreshGlobal(null, null, null, null);
+            State state = ensure(command.txnId());
+            if (!command.status().hasBeen(PreApplied))
+                state.recordBlocking(command.txnId(), Apply, command.someRoutingKeys());
+            state.local().durableGlobal();
+            state.global().durableGlobal(node, command, persistedOn);
         }
 
         @Override
-        public void onFailure(Id from, Throwable failure)
+        public void durable(TxnId txnId, RoutingKeys someKeys, ProgressShard shard)
```

Comment:
This is called durable but doesn't interact with disseminate state and doesn't really seem to be much about durability.

It is also unconditionally recording blocking here, and it's a little unclear how it is blocked by virtue of its own durability.

Reply:
We've been informed that the transaction is durable, i.e. a majority of replicas at all shards have recorded its outcome. So we can (and should) fetch the outcome locally - hence marking blocking, which is all that really does. Logically, our local state being up-to-date is now "blocked" on this